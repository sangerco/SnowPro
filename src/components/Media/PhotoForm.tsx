import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { RootState, AppDispatch } from "../../redux/store";
import { NewPhotoData, createPhoto } from "../../redux/slices/mediaSlices";
import {
  Form,
  Button,
  Dropdown,
  DropdownItemProps,
  Container,
  Divider,
} from "semantic-ui-react";

const PhotoForm: React.FC = () => {
  const auth = useSelector((state: RootState) => state.auth);
  const userId = auth.data?.id;
  const dispatch = useDispatch<AppDispatch>();

  const initialState: NewPhotoData = {
    userId: userId ?? "",
    link: "",
    about: "",
  };

  const [formData, setFormData] = useState(initialState);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    dispatch(createPhoto(formData));

    setFormData(initialState);
  };

  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  return (
    <Container fluid>
      <Form onSubmit={handleSubmit}>
        <Form.Field>
          <label>Photo Link</label>
          <input
            type="text"
            name="link"
            onChange={handleChange}
            value={formData.link}
          />
        </Form.Field>
        <Form.Field>
          <label>About</label>
          <input
            type="text"
            name="about"
            onChange={handleChange}
            value={formData.about}
          />
        </Form.Field>
        <Divider />
        <Divider />
      </Form>
    </Container>
  );
};

export default PhotoForm;
