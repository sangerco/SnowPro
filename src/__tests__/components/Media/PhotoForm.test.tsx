/* eslint-disable testing-library/no-node-access */
import React from "react";
import PhotoForm from "../../../components/Media/PhotoForm";
import nock from "nock";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
describe("test Photo Form", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    media: {
      photo: null,
      photos: [],
      video: null,
      videos: [],
      media: null,
      loading: false,
      error: null,
    },
  };
  const store = mockStore(initialState);
  it("renders with crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <PhotoForm />
        </BrowserRouter>
      </Provider>
    );
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <PhotoForm />
        </BrowserRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render a form", async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <PhotoForm />
        </BrowserRouter>
      </Provider>
    );
    const photoLink = screen.getByLabelText("Photo Link");
    const about = screen.getByLabelText("About");
    const photoForm = screen.getByTestId("photo-form");
    expect(photoLink).toBeInTheDocument();
    expect(about).toBeInTheDocument();
    expect(photoForm).toBeInTheDocument();
  });

  it("should submit the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .post("/api/photos")
      .reply(201, {
        photo: {
          id: "photo-1",
          userId: "testuser",
          link: "https://www.testlinks.com/image.jpg",
          about: "test photo about",
          createdAt: "test created date",
        },
      });
    render(
      <Provider store={store}>
        <BrowserRouter>
          <PhotoForm />
        </BrowserRouter>
      </Provider>
    );
    expect(nock.pendingMocks().length).toBe(1);

    const photoLink = screen.getByLabelText("Photo Link");
    const photoAbout = screen.getByLabelText("About");

    fireEvent.change(photoLink, {
      target: { value: "https://www.testlinks.com/image2.jpg" },
    });
    fireEvent.change(photoAbout, { target: { value: "test photo about" } });

    const photoForm = await screen.findByTestId("photo-form");
    fireEvent.submit(photoForm);

    await waitFor(() => {
      expect(nock.isDone()).toBe(true);
    });
  });
});
