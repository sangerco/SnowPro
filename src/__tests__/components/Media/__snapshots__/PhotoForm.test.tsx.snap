// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Photo Form matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui fluid container"
  >
    <form
      class="ui form"
      data-testid="photo-form"
    >
      <div
        class="field"
      >
        <label
          for="photoLink"
        >
          Photo Link
        </label>
        <input
          id="photoLink"
          name="link"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          for="about"
        >
          About
        </label>
        <input
          id="about"
          name="about"
          type="text"
          value=""
        />
      </div>
      <div
        class="ui divider"
      />
      <button
        class="ui green small button"
        type="submit"
      >
        Send it.
      </button>
      <div
        class="ui divider"
      />
    </form>
  </div>
</DocumentFragment>
`;
